### PR TITLE
murdock: implement test result caching

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This PR adds simple test result caching to Murdock.
The goal is to speed up PR builds with "CI: run tests" set by skipping test applications that have not changed (and would thus have the same result).

Whenever a test is about to be executed on hardware, a hash of the application's binary and all of it's tests is generated. That hash is then looked up in a Redis instance. If the hash exists, the test is skipped.
After a successful test run, the hash is inserted into Redis.
Negative results are never cached.
The cache is not consulted (all test are run) should the label "CI: disable test cache" be set on a PR.
The caching logic is integrated in a way that if anything regarding caching fails (e.g., Redis down), the cache is simply skipped.

This is marked RFC as I'd like some feedback on the approach and its implementation.
There are some open questions:
- this will not catch changes to testrunner. Is there an easy workaround, or is that acceptable?
- this should probably be disabled for the nightlies?

It is also marked WIP as there's ~~no Redis instance running on the Murdock master, yet.~~ (resolved)
*edit* ~~It is also missing on the workers.~~ (resolved)

### Testing procedure

- run tests on this PR
- run tests again. all tests should be skipped. 
- set "CI: disable test cache"
- run tests again. all tests should be executed again

- make a test fail
- run tests again twice, make sure the failed test is re-run every time

- change a test or a tested binary. run tests again, making sure the test is executed

### CODE REVIEW

File: .murdock [hunk 1]
```diff
@@ -131,6 +136,22 @@ print_worker() {
...
+test_hash_calc() {
+    local appdir=$1
+    local bindir=$2
+
+    md5sum $appdir/tests/* $bindir/*.elf | md5sum | cut -f1 -d' '
+}
+
+test_cache_get() {
+    test "${ENABLE_TEST_CACHE}" = "1" || return 1
+    test -n "$(redis-cli get $1)" > /dev/null
```
Comment (miri64) [inline .murdock]: Any reason why Redis is added as a dependency here, while previously we just used the FS as a "database"?

File: .murdock [hunk 2]
```diff
@@ -131,6 +136,22 @@ print_worker() {
         echo "-- running on worker ${DWQ_WORKER} thread ${DWQ_WORKER_THREAD}, build number $DWQ_WORKER_BUILDNUM."
 }
 
+test_hash_calc() {
+    local appdir=$1
+    local bindir=$2
+
+    md5sum $appdir/tests/* $bindir/*.elf | md5sum | cut -f1 -d' '
```
Comment (miri64) [inline .murdock]: Why MD5?

File: .murdock (hunk 1 above)
Comment (miri64) [inline .murdock]: Any reference to the documentation of the added requirement?

**Comment by kaspar030** [comment]
Thanks for your feedback @miri64, those are the kind of questions I think we need to iron out.

> IMHO this should rather go to `RIOT-OS/murdock-scripts` or even `kaspar030/murdock`. 

I thought about it. Reasons for putting it here:

1. `RIOT-OS/murdock-scripts` doesn't know anything about the files involved. It can only determine whether a given test was successful or not. There's no way of hashing the involved files (as they're on the worker). It also doesn't decide which jobs to run.

2. kaspar030/murdock is even further away from the action. All that it does for building is starting "build.sh" and waiting for it to return 0 or 1.

> Putting a new caching mechanism into something that is just _supposed_ to be a project dispatcher script (I know that in some way it isn't, but that's just poor design decisions) seems wrong to me.

What's a project dispatcher script?
In our distributed deployment, the job of './.murdock' is to map RIOT's build process onto Murdock's idea of distributing jobs. IMO, this "prerequisites didn't change, so don't run the test" should be handled by the build system, but it isn't. At that point, why not filter jobs in this script?

File: .murdock (hunk 1 above)
Comment (kaspar030) [reply .murdock]: No. Agreed, some lines somewhere describing the mechanism are needed.

But this is not a hard requirement. Apart from a couble of "redis-cli not found" error messages, nothing of this breaks a build. The cache just won't work if anything is missing.

Comment (miri64) [reply .murdock]: When I wrote this comment, I didn't know Redis can also be setup as a LRU cache (this should also be mentioned in the doc I asked for in https://github.com/RIOT-OS/RIOT/pull/10576#discussion_r240048996 ;-)). I only knew it as key-value store.

File: .murdock (hunk 2 above)
Comment (miri64) [reply .murdock]: > Any reason not to use md5?

There are more advanced hashing algorithms, that have a [least likely chance of collision](https://en.wikipedia.org/wiki/SHA-1#Comparison_of_SHA_functions). The question is more, why would you still use MD5 when e.g. SHA1 (while also not very safe anymore in that regard) is as accessible ;-).

File: .murdock (hunk 1 above)
Comment (miri64) [reply .murdock]: > But this is not a hard requirement. Apart from a couble of "redis-cli not found" error messages, nothing of this breaks a build. The cache just won't work if anything is missing.

I don't like having publicly running systems just run into error cases. Then rather have it documented somewhere and skipped if it isn't installed:

```suggestion
    test -n "$(command -v redis-cli > /dev/null && redis-cli get $1)" > /dev/null
```

Comment (kaspar030) [reply .murdock]: +1. Is the ">/dev/null" necessary for command? On my arch linux it just has empty output when a command is not found.

Comment (miri64) [reply .murdock]: Yepp, but it yields an output *if* the command is found ;-).

File: .murdock (hunk 2 above)
Comment (kaspar030) [reply .murdock]: Least likely chance when under attack. Anyhow, I don't care, will change.

File: .murdock (hunk 1 above)
Comment (kaspar030) [reply .murdock]: Ah right! I read that as ```test -n $(command -v redis-cli) && test -n $(redis-cli get...)```

**Comment by smlng** [comment]
@kaspar030 is there an update to the riot docker image for Murdock-slaves already merged? Then I would first rebuild the k8 image and apply that to the workers.

**Comment by cladmi** [comment]
I agree with the always write to cache, and not read if a condition is set.
Nightlies should indeed not take anything from the cache.

However I would change one thing in the process.
If for any reason, a false positive ends up in the cache, because testrunner was updated or any test scripts dependency that are not taken into account, and no-one re-triggers a build with `Ci: no cache` for the same hash it will stay in the cache for ever.

I would prefer that, before nightly, to flush the cache, and have nightly give a fresh reference cache for the day. This would somehow ensure that a outdated value is garbage collected after some time. Or have a ttl for cache entries which would have a similar behavior but without a deterministic expiration moment from a user perspective.
And during the day after, it means that only the test that changed compared to nightly will be run once which make sense.

Or even maybe only allow nightly to write to the cache. Sure if you change something in core, the test would be run at each PR update but may not be that bad.

We know we will lose somehow the board cpu id dependent outputs like happened in `tests/trickle` but it is also not the goal of the distributed testing in CI and was not deterministic anyway.

**Comment by kaspar030** [comment]
> is there an update to the riot docker image for Murdock-slaves already merged? Then I would first rebuild the k8 image and apply that to the workers.

Yes, current riotdocker:latest includes the redis client!

**Comment by kaspar030** [comment]
> However I would change one thing in the process.
> If for any reason, a false positive ends up in the cache, because testrunner was updated or any test scripts dependency that are not taken into account, and no-one re-triggers a build with `Ci: no cache` for the same hash it will stay in the cache for ever.

Good point.

> I would prefer that, before nightly, to flush the cache

This seems like a good idea.

> Or even maybe only allow nightly to write to the cache. Sure if you change something in core, the test would be run at each PR update but may not be that bad.

Only nightly writing to the cache would mean that PR build time would deteriorate throughout the day, with master moving ahead of the last nightly state. Also, many re-builds are for fixups of unrelated stuff (typos, doxygen, ...), which would be nice to have cached.

We'd also need a documented method for manually flushing the cache.

**Comment by smlng** [comment]
I updated the k8 workers at HAW with the latest image, which should include `redis`.

Otherwise, this seems to work around the more general problem that the RIOT build and test system cannot handle (module) dependencies to finally allow for selected build and tests, i.e. not doing everything all over again all times.

**Comment by miri64** [comment]
Ping? Let's not die this PR ;-).

File: .murdock (hunk 2 above)
Comment (kaspar030) [reply .murdock]: >  SHA1

using sha1 now

Comment (kaspar030) [reply .murdock]: > Could creating a file with the result of `sha1sum $appdir/tests/* $bindir/*.elf` be moved to the riot makefile.

done

**Comment by kaspar030** [comment]
@smlng some of the haw-icc workers cannot connect to Redis. See e.g., [this](https://ci.riot-os.org/RIOT-OS/RIOT/10576/26998bb3ed1c31d9078f8694cd148d17e2d321dc/output/compile/tests/pkg_monocypher/samr21-xpro:llvm.txt) compile output.

Could you take a look?
Are the kubernetes workers also using autossh? Maybe they're missing the autossh change from https://github.com/kaspar030/riotdocker/commit/ed8a162bff322ada91a9d2f1d8ec424214454d7f.

**Comment by smlng** [comment]
@kaspar030 I think they do not have the update in `murdock-slave.sh`, I'll check ...

**Comment by smlng** [comment]
@kaspar030 updated all workers in ICC, should be good now

**Comment by miri64** [comment]
Ping @kaspar030?

**Comment by kaspar030** [comment]
> Ping @kaspar030?

not WIP anymore, and I think the discussion was conclusive. so ... ready for review.

**Comment by kaspar030** [comment]
*ping* this will speed up PR with enabled tests a lot! @smlng maybe you could take another look?

**Comment by cladmi** [comment]
I would still like a way to say "run this not from the cache" because the pull request is changing the flasher, updating `testrunner` or anything. Even if it is just changing a file somewhere to put a "temporary commit" but I would need a way to handle this.
I do not want to have to fix master on the next day because something could not be tested.

If for any reason, the flasher ends up being not working or a testrunner configuration change, it would not be run on the CI for a pull request.

**Comment by kaspar030** [comment]
> I would still like a way to say "run this not from the cache" because the pull request is changing the flasher, updating `testrunner` or anything.

Like a label "CI: disable test cache"?

**Comment by cladmi** [comment]
> > I would still like a way to say "run this not from the cache" because the pull request is changing the flasher, updating `testrunner` or anything.
> 
> Like a label "CI: disable test cache"?

Sorry I missed it was added already as I was not seeing in the labels ><
Could you add the label with the description you want so we can test it is working?

**Comment by kaspar030** [comment]
> Could you add the label with the description you want so we can test it is working?

done.

**Comment by kaspar030** [comment]
I suggest running it once with tests, then again with "disable cache" to see it is re-building, then again without "disable cache" and check that it actually skipped some builds.

**Comment by smlng** [comment]
@kaspar030 is anything missing on the Murdock (master + slaves) side? Or is it just an ACK and merge of this PR? As said I updated the k8s workers in the HAW cloud, I hope they support this now?

**Comment by kaspar030** [comment]
> @kaspar030 is anything missing on the Murdock (master + slaves) side? Or is it just an ACK and merge of this PR? As said I updated the k8s workers in the HAW cloud, I hope they support this now?

Everything should be fine. I've re-build this PR with caching enabled a couple of times and no worker showed problems.

Just ACK & merge missing. ;)

**Comment by kaspar030** [comment]
I've triggered a rebuild. Last build took 37min with all tests run on the hardware (as the cache was empty since last nightly).

**Comment by kaspar030** [comment]
Well, with caching enabled, the build time was ~12min. The cache is now empty (from last nightly) so I'll rebuild again with cache enabled (so it gets populated), then test again with cache disabled via github flag.

**Comment by kaspar030** [comment]
The build with "CI: disable test cache" re-executed all tests. IMO this is good to go.

**Comment by kaspar030** [comment]
- squashed

**Comment by kaspar030** [comment]
- rebased

**Comment by kaspar030** [comment]
@smlng I cringe everytime I have to wait for the same tests to finish again, with the murdock queue getting longer. How can I bribe you into giving this another look? :wink:

File: Makefile.include [hunk 3]
```diff
@@ -580,6 +580,15 @@ test: $(TEST_DEPS)
 test/available:
 	$(Q)test -n "$(strip $(TESTS))"
 
+# this target only makes sense if an ELFFILE is actually created, thus guard by
+# RIOTNOLINE="".
```
Comment (cladmi) [inline Makefile.include]:
```suggestion
# RIOTNOLINK="".
```

Comment (kaspar030) [reply Makefile.include]: fixed

**Comment by kaspar030** [comment]
&go. Thanks!

**Pull request merged**